Implements Bonacich power centrality (networkx#7078)

hehao98 opened the pull request:
This PR is a fairly straightforward implementation of Bonacich power centrality (networkx#7078), following what is implemented in R `igraph`, but additionally considers edge weights (this is really straightforward, didn't understand why the R implementation does not consider this). I have checked that the R `igraph` implementation will return the same or very similar results in all the tests on non-weighted graphs. 

This is my first time submitting a PR to `networkx` so please let me know if I need to change or add anything in this PR.

peijenburg commented:
Thanks for this contribution! It surely looks useful for networkx.
I'm by far an expert on this metric, but still have some suggestions for you:
- You mention that this centrality is a generalisation of the eigenvector centrality. Please add in which case the Bonacich centrality simplifies to the eigenvector centrality.
- Why do you refer to many other centralities under `See also` ? I understand you refer to the eigenvector centrality, but I think the others can be omitted.
- Is the algorithm implemented/meaningful for directed graphs? Please add a test for it, regardless of whether it exists or not.
- The tests only verify the numbers up to 10^-4. Perhaps you can make this a little smaller?
- I'm not a great fan of your usage of `normalised`. Both options `True` and `False` apply a normalisation, with the only difference being a normalisation in the `L_1` or `L_2` norm. I don't know if this parameter originates from theory, but from a user perspective, this parameter is not very intuitive.

hehao98 commented:
Thank you for your comment. Regarding your suggestions:
* I have added the case when the Bonacich centrality simplifies to the eigenvector centrality in the documentation.
* I added them because they are a family of centrality measures that compute the centrality of one node based on the centrality of its neighbors and I was thinking that a reader might need to refer to them when choosing a centrality measure in this family.
* Yes, the algorithm is also implemented for directed graphs (please see test cases `test_directed_P3`, `test_directed_S3`, `test_directed_C3`, and `test_cook_1d_directed`. From [the `igraph` documentation](https://search.r-project.org/CRAN/refmans/igraph/html/power_centrality.html) it seems that it is debated whether the directed case is meaningful. But, directed graphs are used in the original publication (Bonacich, P. (1987)) (replicated in `test_cook_1d_directed`), so I think it would be useful at least in some cases.
* Yes, I have modified the numbers to 10^-7 and the tests still pass.
* This parameter is directly ported from the [the `igraph` version](https://search.r-project.org/CRAN/refmans/igraph/html/power_centrality.html) but I also think it is confusing. The default normalization method (`normalized = False`) is as described in (Bonacich, P. (1987)) but I am personally not aware of any use case for the normalization method in `normalized = False`. If people agree on it, I can remove this parameter and only use the normalization method in (Bonacich, P. (1987))
